pkg/api/kubegraph/analysis/hpa.go

```diff
@@ -0,0 +1,155 @@
+package analysis
+
+import (
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	graphapi "github.com/gonum/graph"
+	osgraph "github.com/openshift/origin/pkg/api/graph"
+	"github.com/openshift/origin/pkg/api/kubegraph"
+	kubenodes "github.com/openshift/origin/pkg/api/kubegraph/nodes"
+	deploygraph "github.com/openshift/origin/pkg/deploy/graph"
+	deploynodes "github.com/openshift/origin/pkg/deploy/graph/nodes"
+)
+
+const (
+	HPAMissingCPUTargetError      = "HPAMissingCPUTarget"
+	HPAOverlappingScaleRefWarning = "HPAOverlappingScaleRef"
+)
+
+// FindHPASpecsMissingCPUTargets scans the graph in search of HorizontalPodAutoscalers that are missing a CPU utilization target.
+// As of right now, the only metric that HPAs can use to scale pods is the CPU utilization, so if a HPA is missing this target it
+// is effectively useless.
+func FindHPASpecsMissingCPUTargets(graph osgraph.Graph, namer osgraph.Namer) []osgraph.Marker {
+	markers := []osgraph.Marker{}
+
+	for _, uncastNode := range graph.NodesByKind(kubenodes.HorizontalPodAutoscalerNodeKind) {
+		node := uncastNode.(*kubenodes.HorizontalPodAutoscalerNode)
+
+		hpa, ok := node.Object().(*extensions.HorizontalPodAutoscaler)
+		if !ok {
+			continue
+		}
+
+		if hpa.Spec.CPUUtilization == nil {
+			markers = append(markers, osgraph.Marker{
+				Node:     node,
+				Severity: osgraph.ErrorSeverity,
+				Key:      HPAMissingCPUTargetError,
+				Message:  fmt.Sprintf("%s is missing a CPU utilization target", namer.ResourceName(node)),
+			})
+		}
+	}
+
+	return markers
+}
+
+// FindOverlappingHPAs scans the graph in search of HorizontalPodAutoscalers that are attempting to scale the same set of pods.
+// This can occur in two ways:
+//   - 1. label selectors for two ReplicationControllers/DeploymentConfigs/etc overlap
+//   - 2. multiple HorizontalPodAutoscalers are attempting to scale the same ReplicationController/DeploymentConfig/etc
+// Case 1 is handled by deconflicting the area of influence of ReplicationControllers/DeploymentConfigs/etc, and therefore we
+// can assume that it will be handled before this step. Therefore, we are only concerned with finding HPAs that are trying to
+// scale the same resources.
+func FindOverlappingHPAs(graph osgraph.Graph, namer osgraph.Namer) []osgraph.Marker {
+	markers := &markerPool{}
+
+	nodeFilterFunction := osgraph.NodesOfKind(
+		kubenodes.HorizontalPodAutoscalerNodeKind,
+		kubenodes.ReplicationControllerNodeKind,
+		deploynodes.DeploymentConfigNodeKind,
+	)
+	edgeFilterFunction := osgraph.EdgesOfKind(
+		kubegraph.ScalingReplicationControllerEdgeKind,
+		kubegraph.ScalingDeploymentConfigEdgeKind,
+		deploygraph.DeploymentEdgeKind,
+	)
+
+	hpaSubGraph := graph.Subgraph(nodeFilterFunction, edgeFilterFunction)
```

Comment:
Because now you follow links in every direction and for every edge separately which is tedious. When we port replica sets, you will need to come back to this code. Compare `generateOverlappingHPAMarkers` to:
```go
for _, edge := range hpaSubGraph.Edges() {
    AddReversedEdge(g, edge.Head(), edge.Tail(), nil)
}
for _, cycle := range topo.CyclesIn(g) {
    // look for more than one hpa in cycle
}
```
:)